src/main/java/com/linecorp/armeria/server/grpc/ArmeriaMessageReader.java

```diff
@@ -0,0 +1,59 @@
+package com.linecorp.armeria.server.grpc;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.http.HttpData;
+import com.linecorp.armeria.common.http.HttpHeaders;
+import com.linecorp.armeria.common.http.HttpObject;
+import com.linecorp.armeria.server.grpc.ArmeriaGrpcServerStream.TransportState;
+
+import io.grpc.Status;
+import io.grpc.internal.ReadableBuffers;
+
+/**
+ * A {@link Subscriber} to read request data and pass it to a GRPC {@link TransportState}
+ * for processing. GRPC code will then handle deframing, decompressing, etc.
+ */
+public class ArmeriaMessageReader implements Subscriber<HttpObject> {
+
+    private final ArmeriaGrpcServerStream.TransportState transportState;
+
+    private Subscription subscription;
+
+    public ArmeriaMessageReader(TransportState transportState) {
+        this.transportState = transportState;
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        this.subscription = subscription;
+        subscription.request(1);
+    }
+
+    @Override
+    public void onNext(HttpObject obj) {
+        if (obj instanceof HttpHeaders) {
+            // Worry about this if armeria ever implements a grpc client.
+            throw new IllegalStateException("Trailing headers read, which can't happen in a server.");
```

Comment:
The behavior is actually that GRPC clients won't ever send them. I've reworded this a bit and actually close the stream properly instead of raising an exception.